### ISSUE TITLE
Integrate logout-from-globus into logout flow

### ISSUE THREAD

**Issue by karlcz** [informatics-isi-edu/chaise#255]
In the globus_auth branch of webauthn, now deployed on isrd-dev and synapse-dev VMs, there is some support to find a Globus logout page where the user can end terminate their upstream authentication session (independent of the session we track in webauthn). 
We should make our chaise logout flow go through the following:
1. DELETE /ermrest/authn/session  (non-interactive)
2. Send browser to Globus logout page. Talk to @ljpearlman about the details. There should be a way to send a referrer or similar link so that the logout page can send the user to the rest of our flow.
3. Send browser to our own post-logout page or back to application?
We need to test the UX and see whether we need our own post-logout page or whether the Globus logout page covers that problem already (in which case having two landing pages in a row might just be annoying).
Chaise should be updated to detect at runtime whether this new webauthn and globus logout option is available, and fall back to the old behavior when encountering the older webauthn logout (such as is still deployed on rbk-dev). That will make the upgrade transitions on each server easier to manage.

**Comment by hongsudt** [comment]
Related to #208.

**Comment by svoinea** [comment]
Updated the chaise code to handle both the old and new webauthn.
The new webauthn is detected by the presence of the client.display_name' attribute.
For the new webauthn, the logout, after deleting the ermrest session, redirects to the Globus logout.

The URL used by chaise for the Globus logout is: https://www.globus.org/app/logout which is a page having a "Log out" link. You have to click it. You will be directed to the Globus https://www.globusid.org/ page.

We need to identify a way such the Globus logout link should redirect to our web application. So far, trying to add it in the URL a referrer or redirect_uri parameter, didn't work (i.e. it was ignored).

Globus is generating a URL like:

https://auth.globus.org/v2/web/logout?client_id=89ba3e72-768f-4ddb-952d-e0bb7305e2c7&client_name=globus_webapp&redirect_uri=https%3A%2F%2Fwww.globus.org%2Fapp&redirect_name=Globus%20Web%20App%20log%20in

**Comment by hongsudt** [comment]
To test new login/logout code:
- test that old globus still work on rbk-dev.isi.edu
- test new globus on dev-isrd.isi.edu

**Comment by hongsudt** [comment]
@svoinea The logout page on dev-rbk doesn't have the same page style as the main rbk.. The message is also confusing.. 

* Need to use the same banner as the rest of the app
* In the text, can we replace "web app" with the project name e.g. (RBK, GPCR, etc.). If not, I will leave that out and just say "Logged Out"  
* Change the last line to "If you want to continue using our data browser, click here." Here should be a link to the data browser (e.g. <hostname>/data, not the main page.
* @crisaless We probably need help with the styling here

![screenshot from 2016-03-24 12-06-39](https://cloud.githubusercontent.com/assets/414732/14028449/d491325c-f1b9-11e5-9a97-646de606b205.png)

**Comment by crisaless** [comment]
It just needs the javascript hooks to the custom css config - Serban has done this for search/index.html.in and record/index.html.in so this just needs the same thing.

**Comment by hongsudt** [comment]
The flow for the new globus authen and logout works. The problem is it is confusing to our users!

**Comment by svoinea** [comment]
Fixed in the commit e8c155c4c5ab3d35da12e78e320d0ac57bf2fed1.

The chaise-config.js file might have a "dataBrowser" parameter to indicate the URL to continue to the data search page. By default, it is "/chaise/search". 

It is important this parameter to be present in the applications that return a "Page Not Found" for the "/chaise/search" URL. It's the Facebase case, where you should have the value  "/data" for the "dataBrowser" parameter.

**Comment by ljpearlman** [comment]
The URL that's being redirected to in logout isn't quite right. The DELETE of /session returns a JSON object that looks like this:
```json
{"logout_url": "https://auth.globus.org/v2/web/logout?redirect_uri=https%3A%2F%2Fwebauthn-dev.isi.edu%2Fchaise%2Flogout&redirect_name=Chaise+logout+page"}
```

That's the URL the user should be redirected to. See https://github.com/informatics-isi-edu/webauthn/blob/globus_auth/doc/rest_api.md for more details.

**Comment by svoinea** [comment]
Fixed the chaise code based on the result returned by DELETE /session.

The Globus logout page looks something like:

    Globus ID — Log out

Continue to https://dev.isrd.isi.edu/chaise/logout 

Seems to me that you need first to click the Globus "Log out" link (which will delete the Globus cookies and will open a Globus login page on a new tab), and then the "https://dev.isrd.isi.edu/chaise/logout" link.

**Comment by ljpearlman** [comment]
Looks good. BTW, I added a globus_auth_logout_redirect_name parameter to the ermrest_config.json on dev.isrd.isi.edu to show what that looks like. @hongsudt might want to change one or both of globus_auth_logout_redirect_name or globus_auth_logout_redirect_path.

**Comment by carlkesselman** [comment]
Minor nit, instead of “our data browser” perhaps “the data browser” or “the Chaise Data Browser”….

Hmm, we should probably have that branding discussion.

Carl

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Mar 29, 2016, at 4:50 PM, ljpearlman <notifications@github.com<mailto:notifications@github.com>> wrote:


Looks good. BTW, I added a globus_auth_logout_redirect_name parameter to the ermrest_config.json on dev.isrd.isi.edu<http://dev.isrd.isi.edu> to show what that looks like. @hongsudt<https://github.com/hongsudt> might want to change one or both of globus_auth_logout_redirect_name or globus_auth_logout_redirect_path.

**Comment by hongsudt** [comment]
Since we already have a globus logout page, I don't see any value in adding our own chaise logout page with the same content. It adds one more click and more confusion. So I am proposing get rid of the chaise logout page. Then on the globus logout, change the last sentence to 
```
To continue using {project-name} data browser, click here. 
``` 
Here should point to the project data browser landing page. According to @ljpearlman this page has to be a static page that doesn't try to log people in. 

For your convenience, I am captured the messages from globus logut and chaise logout content below. 

globus logout: 
```
Logged out of Globus Auth

You have successfully logged out.

If you are using a shared computer, please ensure you have also logged out of any identity providers used during your session, and shut down this browser to remove any session cookies.

To make sure your account is secure, you may need to log out of the following identity providers:

    Globus ID — Log out

Continue to https://synapse-dev.isrd.isi.edu/chaise/logout
```

and if you click the caise/lgout link, it leads you to 
```
Logged out

You have successfully logged out.

If you are using a shared computer, please ensure you have also logged out of any identity providers used during your session, and shut down this browser to remove any session cookies.

If you want to continue using our data browser, click here.
```

**Comment by svoinea** [comment]
This is already done on dev.isrd.isi.edu.

The Globus logout shows now:

Continue to Chaise

And if you click on Chaise, it goes to the /chaise/search page. Therefore, the Chaise logout page is bypassed. 

What you need to do, is to specify these things in the ermrest_config.json file.

For example, in dev.isrd.isi.edu, we have:

      "globus_auth_logout_redirect_path": "/chaise/search",
      "globus_auth_logout_redirect_name": "Chaise"

The Chaise logout page will still be present in the old webauthn, where the Globus logout page is not invoked.

**Comment by hongsudt** [comment]
Can you update the ermrest_config.json for the following projects? 
* gpcr
```
"globus_auth_logout_redirect_path": "/",  --> Need this to go to root (e.g. staging.gpcrconsortium.org) 
"globus_auth_logout_redirect_name": "GPCR Data Browser"
```
* rbk
```
"globus_auth_logout_redirect_path": "/chaise/search",
"globus_auth_logout_redirect_name": "RBK Data Browser"
```
* facebase
```
"globus_auth_logout_redirect_path": "/data/search",
"globus_auth_logout_redirect_name": "FaceBase Data Browser"
```

**Comment by svoinea** [comment]
Currently, this works as expected on rbk-dev and facebase-dev which are doing hourly updates.

However, gpcr-dev is not doing hourly updates for chaise:

Date: Wed,  6 Apr 2016 14:01:08 -0700 (PDT)

/etc/cron.hourly/update-webstack.sh:

U	record/assets/stylesheets/app.css
Pull is not possible because you have unmerged files.
Please, fix them up in the work tree, and then use 'git add/rm <file>'
as appropriate to mark resolution, or use 'git commit -a'.
Chaise git pull failed_

Therefore, the latest logout can not be tested on gpcr-dev.

**Comment by ljpearlman** [comment]
I just logged into/out of gpcr-dev, and it looks like the logout is working there.

**Comment by karlcz** [comment]
This feature doesn't work right when the webauthn session has timed out before the user clicks logout.  It displays the 404 session not found status inside the page rather than accepting 404 as a reasonable status and processing the logout URL embedded in the 404 response's JSON content.

**Comment by svoinea** [comment]
The "session has timed out before the user clicks logout" has been fixed in the commit 985ef66d22aae02835b7f5353fe0e3f31a7fcfab. The user is redirected now to the URL embedded in the ResponseText.

It can be currently tested on rbk-dev.

**Comment by karlcz** [comment]
OK I manually tested by logging in, removing my session from the DB to emulate an exipration, then clicking logout.  The server responded 404 and the GUI sent us to the /chaise/logout page.
